[PATCH v16] api: packet: redefine packet data sharing with packet references

JannePeltonen opened the pull request:
Current API regarding dynamic packet references is not easy to efficiently support in some ODP implementations with HW accelerated packet output due to the need to update reference counts after TX completion and based on that conditionally free packet buffers. odp_packet_ref() has been implemented only as a full packet copy in all ODP implementation.

Redefine the semantics of odp_packet_ref() and the packets involved in packet data sharing in a way that hopefully makes it simpler to support packet data sharing with restrictions in packet output and with per-packet reference counting instead of per-segment reference counting. The new semantics also enable use cases that were not previously possible.

Specify better the rules regarding packet data layout manipulation operations, allowing private packet data also in the tail of the packets that reference other packets. Pull and truncate operations may involve shared packet data, in which case they just remove that data from the packet, not affecting the packet owning the data. All extend and truncate operations always produce non-shared packet data.

Change odp_packet_has_ref() to indicate whether a packet is referenced by another packet, not whether the packet shares data with another packet. In partivular, odp_packet_ref() returns a packet that shares data with the given packet but is not itself referenced by other packets. Specify that packet data and layout of packets referenced by other packets can not be modified at all.

Introduce pktio capability bits to indicate whether a pktio supports sending packets with shared data without using the dont_free flag. Sending with the dont_free flag is supported with all types of packets.

JannePeltonen committed:
api: packet: improve dynamic and static packet references

Current API regarding dynamic packet references is not easy to
efficiently support in some ODP implementations with HW accelerated
packet output due to the need to update reference counts after TX
completion and based on that conditionally free packet buffers.
odp_packet_ref() has been implemented only as a full packet copy in
all ODP implementations.

Redefine the semantics of odp_packet_ref() and the packets involved
in packet data sharing in a way that hopefully makes it simpler to
support packet data sharing with restrictions in packet output and
with per-packet reference counting instead of per-segment reference
counting. The new semantics also enable use cases that were not
previously possible.

Specify better the rules regarding packet data layout manipulation
operations, allowing private packet data also in the tail of the
packets that reference other packets. Pull and truncate operations may
involve shared packet data, in which case they just remove that data
from the packet, not affecting the packet owning the data. All push
and extend operations always produce non-shared packet data.

Change odp_packet_has_ref() to indicate whether a packet is referenced
by another packet, not whether the packet shares data with another
packet. In particular, odp_packet_ref() returns a packet that shares
data with the given packet but is not itself referenced by other
packets. Specify that packet data and layout of packets referenced by
other packets cannot be modified at all.

Add new odp_packet_is_referencing() to test whether a packet
references another packet but is not a static reference.

Introduce pktio capability bits to indicate whether a pktio supports
sending packets with shared data without using the dont_free flag.
Sending with the dont_free flag is supported with all types of packets.

Introduce TM capability bits that indicate if referencing and
referenced packets and static packet references can be enqueued to TM.

Signed-off-by: Janne Peltonen <janne.peltonen@nokia.com>
Reviewed-by: Petri Savolainen <petri.savolainen@nokia.com>
Reviewed-by: Tuomas Taipale <tuomas.taipale@nokia.com>
Acked-by: Ashwin Sekhar T K <asekhar@marvell.com>

JannePeltonen commented on include/odp/api/spec/packet.h:
```diff
@@ -1120,63 +1120,96 @@ odp_packet_t odp_packet_reassemble(odp_pool_t pool, odp_packet_buf_t pkt_buf[],
+ * Packets returned by this function can be used normally in ODP API functions
+ * unless otherwise noted. There are restrictions in packet output.
+ *
+ * Referenced packets can be used in ODP API functions that treat packet data
```
Changed in v2 so that we talk about referenced packets, not base packets.

JannePeltonen commented on include/odp/api/spec/packet.h:
```diff
@@ -1188,16 +1221,18 @@ odp_packet_t odp_packet_ref_pkt(odp_packet_t pkt, uint32_t offset,
+ * of a static reference it also shares metadata. Packet data and data layout
+ * of packet that have references must be treated as read only.
+ *
+ * Packets created through odp_packet_ref() or odp_packet_ref_hdr() share data
```
Fixed in v2

JannePeltonen commented:
v3: Specified that sending shared packets to TM must obey the restrictions of the egress pktio (but currently TM does not support the dont_free flag, which maybe needs to change).

TuomasTaipale commented on include/odp/api/spec/packet.h:
```diff
@@ -1120,63 +1156,121 @@ odp_packet_t odp_packet_reassemble(odp_pool_t pool, odp_packet_buf_t pkt_buf[],
+ * ODP ensures that the packets or packet segments backing the shared packet
+ * data is put back to the packet pool only after there are no references left.
+ * An ODP implementation may implement this using per-packet or per-segment
+ * reference counting or by other means. See odp_packet_free().
```
Unnecessary "An ODP implementation may implement this using per-packet or per-segment reference counting or by other means."?

TuomasTaipale commented on include/odp/api/spec/packet.h:
```diff
@@ -1120,63 +1156,121 @@ odp_packet_t odp_packet_reassemble(odp_pool_t pool, odp_packet_buf_t pkt_buf[],
- * packet consists metadata and data of the 'hdr' packet, followed by the
- * shared part of packet 'pkt'.
+ * 'pkt' must not be a static reference or a referencing packet but it
+ * may be a referenced packet.  'hdr' must not be a static reference or a
```
Extra whitespace before "'hdr'".

TuomasTaipale commented on include/odp/api/spec/packet_io_types.h:
```diff
@@ -1015,6 +1022,32 @@ typedef struct odp_pktin_vector_capability_t {
+	uint8_t referenced  :1; /**< Referenced packets can be consumed */
+	uint8_t referencing :1; /**< Referencing packets can be consumed */
+
+} odp_pktout_packet_ref_capability_t;
```
Better have the same style used elsewhere in file (spec above the field, newlines after fields):
```C
typedef struct {
	/** Static packet references can be consumed */
	uint8_t static_ref  :1;

	/** Referenced packets can be consumed */
	uint8_t referenced  :1;

	/** Referencing packets can be consumed */ 
	uint8_t referencing :1;

} odp_pktout_packet_ref_capability_t;
```

Same comment for the traffic manager capa structure.

psavol replied:
Argee with Tuomas that we should prefer /** style. I think we have cleaned up /**< comments away over the years, but there are still some left.

JannePeltonen replied:
Changed in v6.

TuomasTaipale commented on include/odp/api/spec/packet_io.h:
```diff
@@ -260,6 +260,10 @@ int odp_pktin_queue(odp_pktio_t pktio, odp_pktin_queue_t queues[], int num);
  * are enqueued. Application cannot dequeue from these queues.
  *
+ * Static references, referencing and referenced packets may be sent without
+ * the dont_free option only if the relevant packet_ref capability is set
```
Refer to the full path of `dont_free`?

JannePeltonen replied:
I am not sure what path you mean but added a reference to odp_packet_free_ctrl_set() in v5.

JannePeltonen commented on include/odp/api/spec/packet_io_types.h:
```diff
@@ -1015,6 +1022,32 @@ typedef struct odp_pktin_vector_capability_t {
+ *  left).
+ *
+ *  If these capabilities are not set, then the respective packet types may be
+ *  sent only if the pktout supports the dont_free option and the option is set
```
added in v5

JannePeltonen commented:
v5: Updated based on review comments. Text clarifications only, no semantic changes.

JannePeltonen commented on include/odp/api/spec/packet.h:
```diff
@@ -1120,63 +1149,120 @@ odp_packet_t odp_packet_reassemble(odp_pool_t pool, odp_packet_buf_t pkt_buf[],
+ *
+ * The packet passed to this function must not reference another packet or
+ * be a static reference (see odp_packet_ref_static()). It can already be
+ * a referenced packet.
```
Moved and changed in v6

JannePeltonen commented on include/odp/api/spec/packet.h:
```diff
@@ -1120,63 +1149,120 @@ odp_packet_t odp_packet_reassemble(odp_pool_t pool, odp_packet_buf_t pkt_buf[],
- *               reference. Must be a unique reference.
+ * @param hdr    Handle of the header packet to be prefixed
  *
  * @return New reference the reference packet
```
Fixed in v6

JannePeltonen committed:
api: packet: do not refer to odp_packet_l2_ptr() as data start

Specification text of odp_packet_head() says that odp_packet_data() and
odp_packet_l2_ptr() can be used to return packet data start address. The
latter is incorrect as there is packet data before the l2_ptr if l2 offset
is nonzero. Remove the incorrect reference to odp_packet_l2_ptr().

Signed-off-by: Janne Peltonen <janne.peltonen@nokia.com>
Reviewed-by: Petri Savolainen <petri.savolainen@nokia.com>
Reviewed-by: Tuomas Taipale <tuomas.taipale@nokia.com>
Acked-by: Ashwin Sekhar T K <asekhar@marvell.com>

psavol reviewed (approved):
Reviewed API commits.

JannePeltonen committed:
validation: packet: make packet ref tests work with the coming API too

Make packet reference tests forward compatible with the following changes
that are coming in the API version by removing or changing the conflicting
code.
- odp_packet_has_ref() returns 0 for packets created using odp_packet_ref()
  and odp_packet_ref_pkt().
- Referenced packets may not be modified
- Creating references of references is not allowed

Also fix a couple of asserts that checked the wrong packet handles.

Signed-off-by: Janne Peltonen <janne.peltonen@nokia.com>
Reviewed-by: Tuomas Taipale <tuomas.taipale@nokia.com>

JannePeltonen committed:
linux-gen: packet: reduce cos field size in packet header

Maximum CoS index is 63 but the field for it in packet header is of type
uint16_t. Change the field to uint8_t to make room for other things
without a need to change the memory layout of the header.

Signed-off-by: Janne Peltonen <janne.peltonen@nokia.com>
Reviewed-by: Tuomas Taipale <tuomas.taipale@nokia.com>

JannePeltonen committed:
linux-gen: packet: implement the improved packet references

Introduce the concept of indirect segments whose data pointers point to
data areas of other segments. Use the existing per-segment reference counts
to keep track of the usage of shared segments.

Implement dynamic packet references, or referencing packets, using indirect
segments. A referencing packet has a chain of segments so that some of
the segments may be indirect. Add special handling for indirect segments in
packet manipulation functions wherever necessary.

This commit tries to keep the implementation simple and leaves optimization
opportunities for the future.

Signed-off-by: Janne Peltonen <janne.peltonen@nokia.com>
Reviewed-by: Tuomas Taipale <tuomas.taipale@nokia.com>

JannePeltonen committed:
linux-gen: pktio: add support for packet references

Add support for packet references to all pktios. Many pktios copy packets
before output processing already. Check for referencing packets and make
a copy if needed in loop, ipc and xdp pktio and in the dpdk pktio in the
zero copy case.

Set the relevant pktio capabilities in the common packet i/o code.

Signed-off-by: Janne Peltonen <janne.peltonen@nokia.com>
Reviewed-by: Tuomas Taipale <tuomas.taipale@nokia.com>

JannePeltonen committed:
linux-gen: crypto: add support for referencing packets

If a referencing packet is passed to crypto, copy it before processing to
avoid writing to the read-only parts of the packet.

Signed-off-by: Janne Peltonen <janne.peltonen@nokia.com>
Reviewed-by: Tuomas Taipale <tuomas.taipale@nokia.com>

JannePeltonen committed:
linux-gen: tm: add support for packet references

If a referencing packet is passed to TM, copy it before processing to
avoid writing to the read-only parts of the packet. Unconditional copying
at TM enqueue time could be avoided by making the copy only when needed
in various marking functions but this commit keeps things simple and
does the copy always. Copying for referenced packets and static packet
references was already in place.

Add capability bits to indicate support for packet references.

Signed-off-by: Janne Peltonen <janne.peltonen@nokia.com>
Reviewed-by: Tuomas Taipale <tuomas.taipale@nokia.com>

JannePeltonen committed:
linux-gen: ipsec: add support for referencing packets

If a referencing packet is passed to ipsec, copy it before processing to
avoid writing to the read-only parts of the packet.

Signed-off-by: Janne Peltonen <janne.peltonen@nokia.com>
Reviewed-by: Tuomas Taipale <tuomas.taipale@nokia.com>

JannePeltonen committed:
validation: pktio: add testing with referencing and referenced packets

Add testing with referencing and referenced packets in addition to normal
packets and static references. Add checking of the related new pktio
capabilities.

Signed-off-by: Janne Peltonen <janne.peltonen@nokia.com>
Reviewed-by: Tuomas Taipale <tuomas.taipale@nokia.com>

JannePeltonen committed:
validation: crypto: add testing with referencing and referenced packets

Add testing with referencing and referenced packets in addition to normal
packets and static references.

Signed-off-by: Janne Peltonen <janne.peltonen@nokia.com>
Reviewed-by: Tuomas Taipale <tuomas.taipale@nokia.com>

JannePeltonen committed:
validation: packet: add validation tests for new packet references

Add new validation tests for many basic packet manipulation API functions
and run the tests using many different kind of test packets, including
normal packets and referencing packets.

Signed-off-by: Janne Peltonen <janne.peltonen@nokia.com>
Reviewed-by: Tuomas Taipale <tuomas.taipale@nokia.com>

JannePeltonen committed:
test: bench_packet: add odp_packet_is_referencing()

Add testing of odp_packet_is_referencing().

Signed-off-by: Janne Peltonen <janne.peltonen@nokia.com>
Reviewed-by: Tuomas Taipale <tuomas.taipale@nokia.com>

JannePeltonen committed:
test: packet_gen: check pktio capa when using static reference TX mode

Check that odp_pktio_capability_t::packet_ref.static_ref capability is
present in the used pktios when static reference TX mode is requested.

Signed-off-by: Janne Peltonen <janne.peltonen@nokia.com>
Reviewed-by: Tuomas Taipale <tuomas.taipale@nokia.com>

JannePeltonen committed:
example: sysinfo: print the new packet ref related pktio capa flags

Print the odp_pktout_packet_ref_capability_t flags for pktios.

Signed-off-by: Janne Peltonen <janne.peltonen@nokia.com>
Reviewed-by: Tuomas Taipale <tuomas.taipale@nokia.com>

MatiasElo commented on platform/linux-generic/odp_packet.c:
```diff
@@ -1538,10 +1648,21 @@ void odp_packet_print(odp_packet_t pkt)
+
+			str_len += _odp_snprint(&seg_str[str_len], max_len - str_len,
+						"        referenced seg: seg_data %p "
+						" seg_len %-6" PRIu32 "ref_cnt %u\n",
```
Space missing (`" PRIu32 "ref_cnt %u\n"`)?

JannePeltonen commented on platform/linux-generic/pktio/socket_xdp.c:
```diff
@@ -976,7 +976,8 @@ static int sock_xdp_send(pktio_entry_t *pktio_entry, int index, const odp_packet
 
-		if (_odp_pool_entry(pkt_hdr->event_hdr.pool) != pool) {
+		if (_odp_pool_entry(pkt_hdr->event_hdr.pool) != pool ||
+		    odp_packet_is_referencing(packets[i])) {
```
I ended up not changing this in v12. I thought the code might actually work (?) with static references which is why I only added the dyn ref check.
